lib/http_aws_sigv4.c

```diff
@@ -120,214 +114,277 @@ CURLcode Curl_output_aws_sigv4(struct Curl_easy *data, bool proxy)
+ */
+  str_to_sign = curl_maprintf("%s4-HMAC-SHA256\n" /* Algorithm */
+                              "%s\n" /* RequestDateTime */
+                              "%s\n" /* CredentialScope */
+                              "%s",  /* HashedCanonicalRequest in hex */
+                              provider0_up,
+                              timestamp,
+                              credential_scope,
+                              sha_hex);
   if(!str_to_sign) {
-    ret = CURLE_OUT_OF_MEMORY;
-    goto free_all;
+    goto fail;
   }
 
-  curl_msnprintf(sk, sizeof(sk) - 1, "%s4%s", up_provider,
-                 data->set.str[STRING_PASSWORD]);
-
-  HMAC_SHA256(sk, strlen(sk), date,
-              strlen(date), tmp_sign0);
-  sha256_to_hex(sha_hex, tmp_sign0, sizeof(sha_hex));
+  secret = curl_maprintf("%s4%s",
+                         provider0_up, data->set.str[STRING_PASSWORD]);
+  if(!secret) {
+    goto fail;
+  }
 
-  HMAC_SHA256(tmp_sign0, sizeof(tmp_sign0), region,
-              strlen(region), tmp_sign1);
-  HMAC_SHA256(tmp_sign1, sizeof(tmp_sign1), api_type,
-              strlen(api_type), tmp_sign0);
-  HMAC_SHA256(tmp_sign0, sizeof(tmp_sign0), request_type,
-              strlen(request_type),
-              tmp_sign1);
-  HMAC_SHA256(tmp_sign1, sizeof(tmp_sign1), str_to_sign,
-              strlen(str_to_sign), tmp_sign0);
+  HMAC_SHA256(secret, strlen(secret),
+              date, strlen(date), tmp_sign0);
+  HMAC_SHA256(tmp_sign0, sizeof(tmp_sign0),
+              region, strlen(region), tmp_sign1);
+  HMAC_SHA256(tmp_sign1, sizeof(tmp_sign1),
+              service, strlen(service), tmp_sign0);
+  HMAC_SHA256(tmp_sign0, sizeof(tmp_sign0),
+              request_type, strlen(request_type), tmp_sign1);
+  HMAC_SHA256(tmp_sign1, sizeof(tmp_sign1),
+              str_to_sign, strlen(str_to_sign), tmp_sign0);
 
   sha256_to_hex(sha_hex, tmp_sign0, sizeof(sha_hex));
 
-  auth = curl_maprintf("Authorization: %s4-HMAC-SHA256 Credential=%s/%s, "
-                       "SignedHeaders=%s, Signature=%s",
-                       up_provider, data->set.str[STRING_USERNAME], cred_scope,
-                       signed_headers, sha_hex);
-  if(!auth) {
-    ret = CURLE_OUT_OF_MEMORY;
-    goto free_all;
+  auth_header = curl_maprintf("Authorization: %s4-HMAC-SHA256 "
+                              "Credential=%s/%s, "
+                              "SignedHeaders=%s, "
+                              "Signature=%s",
+                              provider0_up,
+                              data->set.str[STRING_USERNAME],
+                              credential_scope,
+                              signed_headers,
+                              sha_hex);
+  if(!auth_header) {
+    goto fail;
   }
 
-  curl_msnprintf(date_str, sizeof(date_str), "X-%s-Date: %s",
-                 mid_provider, date_iso);
-  data->set.headers = curl_slist_append(data->set.headers, date_str);
+  data->set.headers = curl_slist_append(data->set.headers, auth_header);
```

Comment:
I realize you didn't actually add this, but the code **must not** assign `data->set.headers`. That's the set of custom headers passed to libcurl from the application and it should not be interfered with by internals. It will make repeated transfers and handle duplications misbehave.

Reply:
I'll try to fix it tomorrow.

Reply:
> I realize you didn't actually add this, but the code **must not** assign `data->set.headers`. That's the set of custom headers passed to libcurl from the application and it should not be interfered with by internals. It will make repeated transfers and handle duplications misbehave.

@bagder, see my fix